packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3256 @@
+        // If we don't have a focus node to ask yet, or it can't be focused yet,
+        // then keep the menu until it gets registered, or something else sets
+        // the menu.
+        _pendingFocusedMenu = value;
+      } else {
+        _pendingFocusedMenu = null;
+        value.focusNode!.requestFocus();
+      }
+    }
+    _manageOverlayEntry();
+    _markMenuDirty();
+  }
+
+  // Creates or removes the overlay entry that contains the stack of all menus.
+  void _manageOverlayEntry() {
+    if (openMenu?.topLevel != null) {
+      if (_overlayEntry == null) {
+        _overlayEntry = OverlayEntry(builder: (BuildContext context) => _MenuStack(this));
+        Navigator.of(menuBarContext).overlay!.insert(_overlayEntry!);
+      }
+    } else {
+      _overlayEntry?.remove();
+      _overlayEntry = null;
+    }
+  }
+
+  void _markMenuDirty() {
+    if (_disposed) {
+      return;
+    }
+    menuSerial += 1;
+    _overlayEntry?.markNeedsBuild();
+    notifyListeners();
+  }
+
+  // Build the node hierarchy based upon the MenuItem hierarchy.
+  void _createMenuTree(List<MenuItem> topLevel) {
+    root.children.clear();
+    _focusNodes.clear();
+    _previousFocus = null;
+    _pendingFocusedMenu = null;
+    for (final MenuItem item in topLevel) {
+      _MenuNode(
+        item: item,
+        parent: root,
+      );
+    }
+    assert(root.children.length == topLevel.length);
+  }
+
+  /// Called when updating a MenuBar widget so that the metadata about that menu
+  /// hierarchy and widget can be kept current.
+  ///
+  /// The context is the MenuBar widget's element.
+  ///
+  /// The topLevel is the list of menu items that are the top level children of
+  /// the MenuBar.
+  void update({
+    required BuildContext context,
+    required List<MenuItem> topLevel,
+    required bool enabled,
+  }) {
+    menuBarContext = context;
+    _createMenuTree(topLevel);
+    if (!_listeningToPointerEvents) {
+      GestureBinding.instance.pointerRouter.addGlobalRoute(_handlePointerEvent);
+      _listeningToPointerEvents = true;
+    }
+
+    this.enabled = enabled;
+
+    // Need to mark dirty in a post frame callback, since this update might
+    // updated part of the tree that isn't in the overlay, but calling this
+    // would request that the overlay be rebuilt.
+    SchedulerBinding.instance.addPostFrameCallback((Duration _) {
+      _markMenuDirty();
+    });
+  }
+
+  // Disconnects this controller from the menu bar, in preparation for disposal,
+  // or when the controller is being swapped out for another one.
+  void disconnect() {
+    _menuBarContext = null;
+    root.children.clear();
+    _focusNodes.clear();
+    _previousFocus = null;
+    _focusBeforeClick = null;
+    _pendingFocusedMenu = null;
+    if (_listeningToPointerEvents) {
+      GestureBinding.instance.pointerRouter.removeGlobalRoute(_handlePointerEvent);
+      _listeningToPointerEvents = false;
+    }
+  }
+
+  @override
+  void dispose() {
```

Comment:
uber-nit: maybe move this more towards the end of the class to organize methods according to the controller life cycle?